### ISSUE TITLE
Zoe example: Wrap a ZCFMint to ensure uniqueness for certain NFTs

### ISSUE THREAD

**Issue by katelynsills** [Agoric/agoric-sdk#552]
# User Story

As a user, if I am buying a ticket for a particular seat at a particular show, I want assurance that the mint cannot simply make more tickets for the same seat and double-book it. 

# Design

Create an example Zoe contract that wraps a ZCFMint and checks for uniqueness before minting. The contract should keep an amount of everything that has been minted so far, and `add` newly minted amounts to this variable.

**Comment by erights** [comment]
If I understand the intended semantics correctly, yes, we should enforce uniqueness.

**Comment by katelynsills** [comment]
I think we would need to add a check before line 473:

https://github.com/Agoric/agoric-sdk/blob/2b27ec3cb4f45b911c13a8c76c66044a85651b57/packages/ERTP/src/issuer.js#L472-L474

The check would make sure that for mathHelpers `set` and `strSet`, `newAmount` does not equal and is not included in any amounts currently in `purseLedger` and `paymentLedger`

**Comment by erights** [comment]
That makes sense, but that code is generic across all kinds of amount math. How would you insert the check in a way that maintains the generality of this code?

**Comment by katelynsills** [comment]
We have access to the mathHelpersName, so if the mathHelpersName is included in `['set', 'strSet']` then we could do this additional check.

**Comment by Chris-Hibbert** [comment]
Rather than putting the code in issuer, I would defer to the amountMath. If the amountMath has a method (`exists`, `isUnique`, suggestions?) then I would call that method to determine if a new value was unique.

A previously unmentioned required part of the semantics: if the value had previously been burned, then you can't create it again.

**Comment by katelynsills** [comment]
> Rather than putting the code in issuer, I would defer to the amountMath. If the amountMath has a method (`exists`, `isUnique`, suggestions?) then I would call that method to determine if a new value was unique.

No, I think this is the wrong approach. AmountMath wouldn't know what has been minted before, because it has nothing to do with the internal mappings of purses/payments in issuer. AmountMath isn't stateful. Code in issuer *is* the right place for this check. 

> A previously unmentioned required part of the semantics: if the value had previously been burned, then you can't create it again.

This is a very good point. I could go either way, but we should decide on what we want to do.

**Comment by Chris-Hibbert** [comment]
Yeah, I agree that AmountMath doesn't remember the state, but I still expect the behavior to be specific to the amountMath. I would look for a way for that code before line 473 to pass appropriate state (outstanding purses and payments plus a burnedAmounts list) to an optional method named something like `alreadyMinted()` in amountMath. 

We don't want issuer to start accreting functionality dedicated to particular amountMath types.

**Comment by katelynsills** [comment]
I see what you mean about outsourcing the functionality to AmountMath, now, and I like that idea. Doing anything otherwise would violate the nice polymorphic design we had set up, so that's a very good point. Sorry for being against it at first.

I think, however, we do not want to pass any purses or payments to amountMath. We can certainly get the values of the purses and payments maps (the values would be amounts) and pass those on to AmountMath. 

What if we had an amountMath method such as `amountMath.canMint(prevAmounts, newAmount)` where `prevAmounts` is an array of amounts gotten by getting the values of the purses and payments maps? Then the logic is up to AmountMath.

**Comment by erights** [comment]
The ledgers (`paymentLedger`, `purseLedger`) are weak stores that wrap weak maps. This means you cannot enumerate the contents of the ledgers. If you could, we could never GC any previous payment or purse.

I should have caught this when commenting on https://github.com/Agoric/agoric-sdk/issues/552#issuecomment-619282260 . The plan there also assumes the ability to enumerate for every purse or payment that's ever been stored in the ledger. We cannot become obligated to do that.

I think that means that, for non-fungible goods in general, we can't fix this. We can't know that we're not being asked to mint something again if we don't remember that we've minted it before. And we cannot in general afford to remember everything we've ever minted.

I recommend closing this as can't-fix won't-fix.

**Comment by katelynsills** [comment]
Ah, good point about the weak maps. Just to throw the idea out there, what if we had a separate variable for amounts that have been minted so far? We could even do a `add` on every mint with the newly minted amount such that the storage of all minted amounts is as small as possible. 

I think we do not want to close this as won't fix because it might really be essential for non-fungible mints to guarantee to their users in many cases.

**Comment by erights** [comment]
I don't think we can or should do this for non-fungible goods *in general*. There's no way to remember that set compactly in general. For our current non-fungible amounts, `add` does no significant compaction.

However, it does make sense for opera tickets (and many similar) and is a nice safeguard. We need some way for opera to use this option without making all other non-fungible goods pay those costs.

**Comment by erights** [comment]
Suggestion: The check is only on the proper use of `mint`. Virtualize `mint` into a checking mint much as @dtribble suggested virtualizing offerHook into a checking offerHook.

**Comment by erights** [comment]
Then, `issuer.js` has zero additional complexity, while those who want the check can still impose it on themselves.

**Comment by Chris-Hibbert** [comment]
> Suggestion: The check is only on the proper use of mint. Virtualize mint into a checking mint much as @dtribble suggested virtualizing offerHook into a checking offerHook.

This seems reasonable.

**Comment by katelynsills** [comment]
Sounds good to me!
